server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java

```diff
@@ -396,8 +398,13 @@ public boolean containsBrokenAnalysis(String field) {
      */
     public SearchLookup lookup() {
         if (this.lookup == null) {
+            SourceLoader loader = newSourceLoader(false);
             SourceProvider sourceProvider = isSourceSynthetic() ? (ctx, doc) -> {
-                throw new IllegalArgumentException("Cannot access source from scripts in synthetic mode");
+                int[] docs = new int[] { doc };
+                LeafStoredFieldLoader leafStoredFieldLoader = (loader.requiredStoredFields().isEmpty())
+                    ? StoredFieldLoader.empty().getLoader(ctx, docs)
+                    : StoredFieldLoader.create(false, loader.requiredStoredFields()).getLoader(ctx, docs);
+                return loader.leaf((ctx == null) ? null : ctx.reader(), docs).source(leafStoredFieldLoader, doc);
```

Comment:
This is going to decompress a whole stored field block on every document, potentially re-doing the same work over and over again.  If you want to enable this then I think we need to look again at the SourceLoader.Synthetic implementation, and indeed the SourceLoader API, as it expects to be able to load a complete set of docIds ahead of time and really isn't designed to handle very large numbers of docs.

Reply:
Thanks Alan. I assume we want to decompress the block once and reuse the values we get for fetching docs later.

Do you have any good pointers on how to make these changes? I'll take a closer look to SourceLoader internals too.

Reply:
Yes, we ideally want to have just a single LeafStoredFieldLoader and SourceLoader.Leaf per segment.  I think you want a new SourceProvider implementation that's tailored specifically for synthetic source, and you'll want to adjust the various implementations of SyntheticFieldLoader.DocValuesLoader to fall back to doc-by-doc loading if the passed in docids array is `null`.

Reply:
Updated, is this closer to what you had in mind? Thanks.